### ISSUE TITLE
Cosine annealing called every epoch instead of every batch

### ISSUE THREAD

**Issue by guybuk** [skorch-dev/skorch#610]
Cosine annealing LR should update the learning rate every batch according to: https://arxiv.org/pdf/1608.03983.pdf

It does not, seeing as the method `skorch.LRScheduler.on_batch_end` is: 
```python
    def on_batch_end(self, net, training, **kwargs):
        if (
                training and
                hasattr(self.lr_scheduler_, 'batch_step') and
                callable(self.lr_scheduler_.batch_step)
        ):
            self.lr_scheduler_.batch_step(self.batch_idx_)

        if TorchCyclicLR and isinstance(self.lr_scheduler_, TorchCyclicLR):
            self.lr_scheduler_.step(self.batch_idx_)

        if training:
            self.batch_idx_ += 1
```
`torch.optim.lr_scheduler.CosineAnnealingLR` is neither `TorchCyclicLR` nor has a `batch_step` call (On that note, I haven't found a single LRScheduler that does have a `batch_step` method).

And on a somewhat unrelated note, it would be ideal for me to log somehow the LR per batch (or epoch, depending on the scheduler). Would it be wise to add a `net.record(...)` / `net.record_batch` call to `skorch.LRScheduler`?

**Comment by BenjaminBossan** [comment]
Thanks for reporting.

I'm not an expert on learning rate schedulers, so please correct me if I'm wrong. From my reading of things, the `CosineAnnealingLR` in pytorch is intended to work on an epoch level. They write:

> Set the learning rate of each parameter group using a cosine annealing schedule, where η\_max​ is set to the initial lr and T_cur​ is the number of epochs since the last restart in SGDR:

[docs](https://pytorch.org/docs/stable/optim.html#torch.optim.lr_scheduler.CosineAnnealingLR)

For me, it reads like this is the intended behavior. I cannot judge if this is the correct behavior according to the paper, but if it isn't, and if my interpretation is correct, you would need to open an issue in pytorch and report this error. I wouldn't go so far as to try to correct a wrong implementation in pytorch on the skorch side.

But again, I might be reading things wrong here, so please correct me if that's the case.

Regarding the missing `batch_step`: I believe you are right, we can remove it. The reason why it existed in the first place was that initially, skorch implemented its own learning rate schedulers as they were not supported by pytorch yet (see [here](https://github.com/skorch-dev/skorch/pull/180/files#diff-33ab920250ece50e2870e0753826449fR286)). Since @thomasjpfan implemented this, could you please confirm that this code can be removed?

> And on a somewhat unrelated note, it would be ideal for me to log somehow the LR per batch (or epoch, depending on the scheduler). Would it be wise to add a `net.record(...)` / `net.record_batch` call to `skorch.LRScheduler`?

Yes, I guess this could be added. Do you want to work on that feature? I wonder if the key should start with `event_` (as e.g. used by the `Checkpoint` callback) so that it isn't logged by default to, say, tensorboard.

**Comment by thomasjpfan** [comment]
> Regarding the missing batch_step

We can remove the `batch_step` code.

**Comment by guybuk** [comment]
@BenjaminBossan I'd have to say it is Pytorch's own documentation that's confusing. At the beginning it says:
> T_cur is the number of epochs since the last restart
And a few lines under that, in the `Parameters` section it reads:
> T_max (python:int) – Maximum number of iterations.
Which under reasonable logic (Since T_cur is divided by T_max)  would mean that T_cur should also be iterations.

In the original paper I have linked above it is also explicitly stated:
> Tcur is updated at each batch iteration.

> Do you want to work on that feature?
Gladly. Might need a bit of hand holding since I've never actually opened a PR in Github before, but I'll do my best to keep that to the minimum.

**Comment by BenjaminBossan** [comment]
> I'd have to say it is Pytorch's own documentation that's confusing.

I agree that the docs are confusing and even after reading the code, I'm not quite clear on how it's supposed to be used. "Iteration" is ambiguous to me, it could mean both "epoch" or "batch". Do you think you could clarify with the pytorch devs (on their forum or create an issue) how `CosineAnnealingLR` is supposed to be used?

In case it's meant to be updated per batch, I think skorch would need to be updated analogously to here:

https://github.com/skorch-dev/skorch/blob/e61f10c18f2f7c6251615c3b413e9f3de78062eb/skorch/callbacks/lr_scheduler.py#L157-L158

> Gladly. Might need a bit of hand holding since I've never actually opened a PR in Github before, but I'll do my best to keep that to the minimum.

Thanks. If you have any questions, just ask them and we'll try to help you. From a technical point of view, you might want to read about creating [forks](https://help.github.com/en/github/getting-started-with-github/fork-a-repo) and then about creating [PRs](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/creating-a-pull-request-from-a-fork).

**Comment by guybuk** [comment]
> "Iteration" is ambiguous to me, it could mean both "epoch" or "batch".

```python
>>> scheduler = CosineAnnealingWarmRestarts(optimizer, T_0, T_mult)
>>> iters = len(dataloader)
>>> for epoch in range(20):
>>>     for i, sample in enumerate(dataloader):
>>>         inputs, labels = sample['inputs'], sample['labels']
>>>         optimizer.zero_grad()
>>>         outputs = net(inputs)
>>>         loss = criterion(outputs, labels)
>>>         loss.backward()
>>>         optimizer.step()
>>>         scheduler.step(epoch + i / iters)
```

Taken from here: https://pytorch.org/docs/master/optim.html#torch.optim.lr_scheduler.CosineAnnealingWarmRestarts 


Since the only difference is the warm restarts I figure this should be enough, but if not I guess I can post a question somewhere

**Comment by BenjaminBossan** [comment]
Yes, for `CosineAnnealingWarmRestarts`, it seems that iterations actually refers to batches. We could change the `LRScheduler` in the way mentioned above so that it works on batch level.

I wonder, however, if this is the best approach. We basically have a fixed set of schedulers that are updated on batch level, all others on epoch level. Would it be better to add an option to `LRScheduler` to control whether epoch or batch level is desired?

**Comment by guybuk** [comment]
> Would it be better to add an option to LRScheduler to control whether epoch or batch level is desired?

I think it would. It's more manageable than adding every new scheduler-by-batch to the `if` statement in `on_batch_end`. Since Pytorch themselves give the freedom to play with it (Regardless of whether the original purpose for each scheduler was to step every epoch or every batch, they all use the same `step` method) I think it's fair skorch does the same.

On that note, looking at this piece of code inside `LRScheduler`:
```python
    def _get_policy_cls(self):
        if isinstance(self.policy, str):
            return getattr(sys.modules[__name__], self.policy)
        return self.policy
```

Is having the ability to accept string values for `policy` actually worth having to import every single scheduler skorch may support? From my personal experience, I always use a class value for `policy` anyway, if only to have an easy way to look at what parameters my scheduler uses through IntelliSense. So if strings are no longer accepted for `policy` it could clean up some code and save some work.

**Comment by BenjaminBossan** [comment]
> I think it would.

I believe it would actually quite a nice feature to have. Do you want to work on it?

> So if strings are no longer accepted for `policy` it could clean up some code and save some work.

I personally would do the same as you do but there may be others who like the ability to pass a string. Removing that possibility would be backwards-incompatible, so unless there is a very strong reason to do it, I would keep the option.

**Comment by guybuk** [comment]
> Do you want to work on it?

Sure.

**Comment by guybuk** [comment]
Since my previous PR got merged, I'll work on this one.

Question: If we add a `step_on` option for `LRScheduler`, would we like the possible values to be `batch_end` & `epoch_end` to align with the currently existing `on_epoch_end` and `on_batch_end`? Or would we prefer simple `step_every='batch'/'epoch'`?

**Comment by BenjaminBossan** [comment]
Thanks for taking this.

I would only choose `batch_end` or `epoch_end` as values if there was any possibility that `batch_start` and `epoch_start` could one day also be considered. My imagination fails me, however, on why that would ever make sense. Hence I would prefer `batch` and `epoch`. From there, we could still eventually transition to a more fine grained control if ever necessary, without breaking backwards compatibility.
